kubetest/extract_test.go

```diff
@@ -83,3 +83,58 @@ func TestGetKube(t *testing.T) {
 		}
 	}
 }
+
+func TestExtractStrategies(t *testing.T) {
+	cases := []struct {
+		option        string
+		expectURL     string
+		expectVersion string
+	}{
+		{
+			"bazel/v1.8.0-alpha.2.899+2c624e590f5670",
+			"",
+			"bazel/v1.8.0-alpha.2.899+2c624e590f5670",
+		},
+		{
+			"bazel/49747/master:b341939d6d3666b119028400a4311cc66da9a542,49747:c4656c3d029e47d03b3d7d9915d79cab72a80852",
+			"",
+			"bazel/49747/master:b341939d6d3666b119028400a4311cc66da9a542,49747:c4656c3d029e47d03b3d7d9915d79cab72a80852",
+		},
+	}
+
+	var gotURL string
+	var gotVersion string
+
+	// getKube is tested independently, so mock it out here so we can test
+	// that different extraction strategies call getKube with the correct
+	// arguments.
+	oldGetKube := getKube
+	defer func() { getKube = oldGetKube }()
+	getKube = func(url, version string) error {
+		gotURL = url
+		gotVersion = version
+		// This is needed or else Extract() will think that getKube failed.
+		os.Mkdir("kubernetes", 0775)
+		return nil
+	}
+
+	for _, tc := range cases {
+		if d, err := ioutil.TempDir("", "extract"); err != nil {
+			t.Fatal(err)
+		} else if err := os.Chdir(d); err != nil {
+			t.Fatal(err)
+		}
+
+		var es extractStrategies
```

Comment:
It doesn't *need* to be, but it was convenient for testing exactly what I cared about.

`extractStrategy` doesn't have a constructor, so I can only create one by manually setting all of its fields correctly. `extractStrategies`, on the other hand, is what takes the option string and translates it into one or more `extractStrategy`s in its [Set() method](https://github.com/kubernetes/test-infra/blob/a97d5b8c0c166bced91f911f66eed523bf55976a/kubetest/extract.go#L68).

If I wanted to test at the `extractStrategy` level, then I would need two separate tests:

1) Make sure that `extractStrategies.Set()` translates my `bazel/...` option string into an `extractStrategy` that has `mode == bazel`
2) Construct an `extractStrategy` with `mode = bazel` and make sure that when I call `Extract()` that it passes the correct parameters to `getKube()`.

I personally don't care about the internal representation used when I pass in a bazel version string, what I actually care about is that when a user specifies a version that looks like `bazel/...`, that the correct version string and URL are passed to `getKube`. Validating any steps in the middle just makes the test more brittle in the face of refactoring for no added value (based on the scenario that I care about asserting works).